### PR TITLE
Added Try/Catch to Process-Handler function in handler.ps1 to verify …

### PR DESCRIPTION
**Pull request by meyeaard**
Added Try/Catch to Process-Handler function in handler.ps1 to verify vCenter session has not expired both generally and specifically for tagging operations as this sub-session times out earlier than the general vCenter PowerCLI session.

Signed-off-by: Aaron Meyer <ameyer@meyertech.net>

## Summary

Please describe your changes and what you plan to introduce as part of this pull request submission. If it fixes a bug or resolves a feature request, be sure to link to the issue numbers following the template below.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [ ] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [ ] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [ ] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

Closes #ISSUE-NUMBER

## Testing Verification

* Short summary of testing (e.g. successfully built and deployed VMware Event Broker Appliance)
* Please include any relevant screenshots and/or output as part of the testing

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

File: examples/knative/powercli/kn-pcli-tag/handler.ps1 [hunk 1]
```diff
@@ -71,6 +71,16 @@ Function Process-Handler {
...
    $vmName = $cloudEventData.Vm.Name
 
+   # Ensure vCenter session is still viable for Tag actions (This expires before the general session)
+   Try {
+      Get-TagCategory -ErrorAction:stop | Out-Null
+      Write-Host "$(Get-Date) - vCenter session verified viable for tagging."
+   } Catch {
+      Write-Host "$(Get-Date) - vCenter session for tagging has expired. Reinitializing..."
+      Disconnect-VIServer -Server * -Confirm:$false -Force:$true
+      Process-Init
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-tag/handler.ps1:81]: What if `Process-Init` throws again? The current behavior would be that the handler returns, the exception would be handled by the server.ps try/catch and return INTERNAL_SERVER_ERROR to the caller (e.g. broker). A broker would generally retry, potentially leading to a (temporary) loop.

Either we accept this (until we return better/typed errors to the server.ps code) or we exit 1 here to crash the function.

**Comment by kremerpatrick** [comment]
What if we specifically catch only VMware.VimAutomation.Sdk.Types.V1.ErrorHandling.VimException.ViServerConnectionException  ? 

This will bypass the potential loop while catching the error that we generally want to catch "Server X is not connected"

**Comment by embano1** [comment]
> What if we specifically catch only VMware.VimAutomation.Sdk.Types.V1.ErrorHandling.VimException.ViServerConnectionException ?

I think there's not much benefit in type assertion here as the only thing we can (should be?) doing for now is exit 1?

**Comment by embano1** [comment]
After some discussions on related issues, I'm in favor of changing `server.ps1` to default to `$ErrorActionPreference = 'Stop'` and have handler commands interrupt/throw an exception which is already caught be the outer server try/catch. Also see discussion in #579.

In case of auth issues, it's actually better to fail fast and not complicate the logic (e.g. by calling init again). The container will `exit 1` and restart automatically. 

Currently (due to an issue in the eventing broker for rabbit), events are always retried (even though we throw 500 currently). Once this is fixed we need to think about dead-letter implementations so events are not lost (and can be manually replied/inspected). An alternative is to implement typed exceptions so the outer try/catch (in server) can distinct between permanent/user/temporary errors and retry accordingly (by responding with a specific HTTP status code).

**Comment by embano1** [comment]
IMHO this PR can be closed since we added https://github.com/vmware-samples/vcenter-event-broker-appliance/pull/579 to automatically catch all exceptions from handler.

**Comment by kremerpatrick** [comment]
Agreed

**Pull request closed without merging**